Pagination

fivitti opened the pull request:
This MR adds pagination solution to our server.
It provide easy to use (I hope), but flexible solution to add pagination.
You have a different ways to use pagination tools which differ in the amount of work to be done. See docstrings.
I apply it to ```obslist``` and ```stations```template.

This pagination is page-based. It means that user may select which "page" want to see.
If page is only one then pagination controls aren't display.

The summary with items count is now generated automatically.

Version from your previous commit s now displayed on all pages without the need for manually include footer.

Related to #21

tomaszmrugalski commented on server/app/controllers/obs.py:
```diff
@@ -1,11 +1,18 @@
         abort(300, description="ID is required")
+    if limit_and_offset is None:
+        limit_and_offset = {
+            'limit': 5,
```
This looks like a default value. It should be clearly marked as such. Also, this should be equal to what is specified in the items_per_page value in satnogs.ini.

fivitti replied:
Fixed. I remove this parameters. It doesn't affect, because None value is rewrite by ```use_pagination``` decorator.

tomaszmrugalski commented on server/app/controllers/obslist.py:
```diff
@@ -1,17 +1,19 @@
 
-    return render_template('obslist.html', obslist=obslist)
\ No newline at end of file
+    return 'obslist.html', dict(obslist=obslist, item_count=observation_count)
```
EOL missing.

tomaszmrugalski commented on server/app/pagination.py:
```diff
@@ -0,0 +1,240 @@
+Pagination helpers. You have three ways to use this tools:
+
+1. Low-level. You should create Paginate class object and use them to calculate
+    all needed parameters. You must manuall retrive page number and pass
```
```suggestion
    all needed parameters. You must manually retrieve page number and pass
```

tomaszmrugalski commented on server/app/pagination.py:
```diff
@@ -0,0 +1,240 @@
+            'page_count': self._page_count(item_count),
+            'items_count': item_count,
+            'items_current': self._page_item_count(item_count, current)
+
```
Extra empty line.  Pls remove

tomaszmrugalski commented on server/app/pagination.py:
```diff
@@ -0,0 +1,240 @@
+    3. Create object of Pagination class
+    4. Calculate limit and offset
+    5. Pass limit and offset to specific repository read method by unpack
+        content by duble asteriks
```
```suggestion
        content by double asterisks
```

tomaszmrugalski commented on server/app/pagination.py:
```diff
@@ -0,0 +1,240 @@
+    >>> {% include 'pagination.html' %}
+    '''
+    return use_kwargs({
+        PARAM_NAME: webargs.fields.Int(missing=1, 
```
Why is it wrapped? What's your preference for max line length?

fivitti replied:
From PEP8, official Python style guide:
>Limit all lines to a maximum of 79 characters.
>
>For flowing long blocks of text with fewer structural restrictions (docstrings or comments), the line length should be limited to 72 characters.

I think that normal lines should have max 80 character, if you have good reason then it may be longer, but avoid line longer then 120 characters.
I often have two windows side-by-side and this limitations are useful for me.

tomaszmrugalski replied:
I dislike overly wrapped text. It causes less code to fit on the screen and you need to scroll more. Anyway, this is just a personal preference, so let's keep the code as it is.

tomaszmrugalski commented on server/app/pagination.py:
```diff
@@ -0,0 +1,240 @@
+'''
```
Is this code imported? What's its license?

fivitti replied:
No. It is totally my job!

tomaszmrugalski replied:
Awesome! Well done then!

tomaszmrugalski commented on server/tests/test_pagination.py:
```diff
@@ -0,0 +1,47 @@
+        paginaton = Pagination(items_per_page)
+
+        limit_and_offset = paginaton.limit_and_offset(1)
+        self.assertDictContainsSubset({
```
The tests are passing, but there's a warning about assertDictContainsSubset being deprecated:

```
tests/test_pagination.py::TestPagination::test_limit_and_offset
  /usr/lib/python3.7/unittest/case.py:1156: DeprecationWarning: assertDictContainsSubset is deprecated
    DeprecationWarning)

-- Docs: https://docs.pytest.org/en/latest/warnings.html
```
That's on ubuntu 19.10 with python 3.7.5.

fivitti replied:
Fixed based on [StackOverflow](https://stackoverflow.com/a/57386339)

tomaszmrugalski replied:
The warning is now gone.

tomaszmrugalski commented on server/templates/pagination.html:
```diff
@@ -0,0 +1,44 @@
+  </nav>
+{% endif %}
+<p class="text-center">
+    Display {{ pagination.items_current }} from {{ pagination.items_count }} items
```
```suggestion
    Displaying {{ pagination.items_current }} out of {{ pagination.items_count }} item(s).
```

tomaszmrugalski commented:
I've tested in on Ubuntu 19.10 with python 3.7.5. Seems to be working fine. Unit-tests pass. Contrary to branch name, there's no searching or filtering. But I assumed it's not part of the PR.

Here's a tip: you can easily "rename" branches.
```
git checkout search-pagginate-filter
git checkout -b paginate
git push -u origin paginate
git branch -D search-pagginate-filter
git push origin :search-pagginate-filter
```

fivitti commented:
I fixed your notes, but I added solution for handle multiple pagination on the same page. 
Please review changes from [this commit](https://github.com/gut-space/satnogs/pull/58/commits/8fcd93c39015b4ebae533982057c6b680496bb61).

tomaszmrugalski commented on server/app/repository.py:
```diff
@@ -148,6 +148,14 @@ def _cursor(self) -> Cursor:
 
+    @use_cursor
+    def count_observations(self) -> int:
+        q = ("SELECT COUNT(*) as count "
```
Is your terminal 40 chars wide? This kind or line wrap looks horrible.

tomaszmrugalski replied:
You mentioned PEP8 saying line wrap at 80 chars. Are we moving to PEP4 now? :rofl:
